src/project/mod.rs

```diff
@@ -0,0 +1,126 @@
+use cerberus::project::ProjectData;
+use std::sync::Arc;
+use std::time::Instant;
+
+use cerberus::registry::{RegistryClient, RegistryError, RegistryHttpClient, RegistryResult};
+use common::metrics::AppMetrics;
+use common::storage::{redis, StorageError};
+
+pub use config::*;
+pub use error::*;
+
+use crate::error::{RpcError, RpcResult};
+use crate::project::metrics::ProjectDataMetrics;
+use crate::project::storage::ProjectStorage;
+use crate::project::storage::{Config as StorageConfig, ProjectDataResult};
+
+mod config;
+mod error;
+
+pub mod metrics;
+pub mod storage;
+
+#[derive(Debug, Clone)]
+pub struct Registry {
+    client: RegistryHttpClient,
+    cache: Option<ProjectStorage>,
+    metrics: ProjectDataMetrics,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum ResponseSource {
+    Cache,
+    Registry,
+}
+
+impl Registry {
+    pub fn new(cfg_registry: &Config, cfg_storage: &StorageConfig) -> RpcResult<Self> {
+        let api_url = &cfg_registry.api_url;
+        let api_auth_token = &cfg_registry.api_auth_token;
+
+        let client = if let (Some(api_url), Some(api_auth_token)) = (api_url, api_auth_token) {
+            RegistryHttpClient::new(api_url, api_auth_token)?
+        } else {
+            return Err(RpcError::InvalidConfiguration(
+                "missing registry api parameters".to_string(),
+            ));
+        };
+
+        let metrics = ProjectDataMetrics::new(&AppMetrics::new(crate::PROXY_METRICS_NAME));
+
+        let cache_addr = cfg_storage.project_data_redis_addr();
+        let cache = match cache_addr {
+            None => None,
+            Some(cache_addr) => {
+                let cache = open_redis(&cache_addr, cfg_storage.redis_max_connections)?;
+
+                Some(ProjectStorage::new(
+                    cache,
+                    cfg_registry.project_data_cache_ttl(),
+                    metrics.clone(),
+                ))
+            }
+        };
```

Comment:
Not sure if this is better, but:
```suggestion
        let cache = cfg_storage
            .project_data_redis_addr()
            .map(|addr| open_redis(&addr, cfg_storage.redis_max_connections))
            .transpose()?
            .map(|conn| {
                ProjectStorage::new(conn, cfg_registry.project_data_cache_ttl(), metrics.clone())
            });
```

Reply:
It looks fantastic, but I think the original communicates intent better, and I'm not sure there's a difference performance-wise.